trillian/integration/ct_integration.go

```diff
@@ -437,80 +501,63 @@ func RunCTIntegrationForLog(cfg *configpb.LogConfig, servers, metricsServers, te
-	}
-	hash := sha256.Sum256(append([]byte{merkletree.LeafPrefix}, leafData...))
-	rsp, err := t.client().GetProofByHash(ctx, hash[:], sthN1.TreeSize)
-	t.stats.done(ctfe.GetProofByHashName, 200)
-	if err != nil {
-		return fmt.Errorf("got GetProofByHash(precertSCT, size=%d)=nil,%v", sthN1.TreeSize, err)
-	}
-	if rsp.LeafIndex != precertIndex {
-		return fmt.Errorf("got GetProofByHash(precertSCT, size=%d).LeafIndex=%d; want %d", sthN1.TreeSize, rsp.LeafIndex, precertIndex)
-	}
-	fmt.Printf("%s: Inclusion proof leaf %d @ %d -> root %d = %x\n", t.prefix, precertIndex, precertSCT.Timestamp, sthN1.TreeSize, rsp.AuditPath)
-	if err := Verifier.VerifyInclusionProof(rsp.LeafIndex, int64(sthN1.TreeSize), rsp.AuditPath, sthN1.SHA256RootHash[:], leafData); err != nil {
-		return fmt.Errorf("got VerifyInclusionProof(%d,%d,...)=%v; want nil", precertIndex, sthN1.TreeSize, err)
-	}
-	if err := t.checkStats(); err != nil {
-		return fmt.Errorf("unexpected stats check: %v", err)
+	if err := t.checkInclusionOfPreCert(ctx, tbs, issuer, precertSCT, sthN1); err != nil {
+		return fmt.Errorf("failed to check inclusion of pre-cert entry: %v", err)
 	}
 
 	// Stage 15: invalid consistency proof
 	if rsp, err := t.client().GetSTHConsistency(ctx, 2, 299); err == nil {
 		return fmt.Errorf("got GetSTHConsistency(2,299)=(%+v,nil); want (nil,_)", rsp)
 	}
-	t.stats.done(ctfe.GetSTHConsistencyName, 400)
+	t.stats.done(ctfe.GetSTHConsistencyName, 500)
 	fmt.Printf("%s: GetSTHConsistency(2,299)=(nil,_)\n", t.prefix)
 
 	// Stage 16: invalid inclusion proof
 	wrong := sha256.Sum256([]byte("simply wrong"))
 	if rsp, err := t.client().GetProofByHash(ctx, wrong[:], sthN1.TreeSize); err == nil {
 		return fmt.Errorf("got GetProofByHash(wrong, size=%d)=(%v,nil); want (nil,_)", sthN1.TreeSize, rsp)
 	}
-	t.stats.done(ctfe.GetProofByHashName, 400)
+	t.stats.done(ctfe.GetProofByHashName, 500)
 	fmt.Printf("%s: GetProofByHash(wrong,%d)=(nil,_)\n", t.prefix, sthN1.TreeSize)
 
+	// Stage 17: build and add a pre-certificate signed by a pre-issuer.
+	preIssuedChain, preTBS, err := makePreIssuedPrecertChain(chain[1], issuer, signer)
+	if err != nil {
+		return fmt.Errorf("failed to build pre-issued pre-certificate: %v", err)
+	}
+	preIssuedCertSCT, err := pool.Next().AddPreChain(ctx, preIssuedChain)
+	stats.done(ctfe.AddPreChainName, 200)
+	if err != nil {
+		return fmt.Errorf("got AddPreChain()=(nil,%v); want (_,nil)", err)
+	}
+	fmt.Printf("%s: Uploaded pre-issued precert to %v log, got SCT(time=%q)\n", t.prefix, precertSCT.SCTVersion, timeFromMS(precertSCT.Timestamp))
+	treeSize++
+	sthN2, err := t.awaitTreeSize(ctx, uint64(treeSize), true, mmd)
+	if err != nil {
+		return fmt.Errorf("AwaitTreeSize(%d)=(nil,%v); want (_,nil)", treeSize, err)
+	}
+	fmt.Printf("%s: Got STH(time=%q, size=%d): roothash=%x\n", t.prefix, timeFromMS(sthN2.Timestamp), sthN2.TreeSize, sthN2.SHA256RootHash)
+
+	// Stage 18: retrieve and check pre-issued pre-cert.
+	preIssuedCertIndex := int64(count + 2)
```

Comment:
We only know these indices because nothing else should be touching the log, right?